### PR TITLE
Llama3-70b: Full Finetune w/CPU offload + fused optimizer

### PR DESCRIPTION
**Pull request by rohan-varma**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
- This PR enables 70B full finetune for Llama3 workload on 8xA100 set up with CPU offload. The configuration offered only works with PyTorch nightly which has recent changes to support `fused=True` for AdamW to utilize vectorized CPU instructions.
- Similar to https://github.com/pytorch/torchtune/pull/802, we use the HF checkpointer and load the model with HF format, with the meta format currently unsupported.
- QPS looks as follows: `1|95|Loss: 0.7184870839118958:   3%|▎         | 95/3251 [22:02<11:53:47, 13.57s/it]` - while slow, this actually seems competitive with 70B training workloads seen in other libraries.
- Appropriate warnings and error checking is added
- This PR also calls `torch.set_num_threads` after set up to ensure that the appropriate # of CPU cores are set for intra-op parallelism. On my box, this results in ~2x speed up compared to not setting it. I got the idea to do this after talking with @msaroufim @janeyx99 

#### Test plan

```
tune run --nproc_per_node 8 full_finetune_distributed --config recipes/configs/llama3/70B_full.yaml metric_logger=torchtune.utils.metric_logging.WandBLogger log_peak_memory_stats=True
```

Memory usage, sits at ~35GB active per GPU, 42 GB reserved - 

image

Loss looks comparable to 70B  model trained w/LoRA (see loss curves https://github.com/pytorch/torchtune/pull/802) - 
image

### CODE REVIEW

File: recipes/configs/llama3/70B_full.yaml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+seed: null
+shuffle: True
+
+# Model Arguments
+model:
+  _component_: torchtune.models.llama3.llama3_70b
+
+checkpointer:
+  _component_: torchtune.utils.FullModelHFCheckpointer
+  checkpoint_dir: /tmp/Meta-Llama-3-70b
```
Comment (rohan-varma) [inline recipes/configs/llama3/70B_full.yaml:37]: need to covert the paths to instruct

File: recipes/full_finetune_distributed.py [hunk 2]
```diff
@@ -563,6 +584,12 @@ def recipe_main(cfg: DictConfig) -> None:
         )
 
     init_process_group(backend="gloo" if cfg.device == "cpu" else "nccl")
+    if cfg.get("fsdp_cpu_offload", False):
+        # Utilize all available CPU cores for intra-op parallelism. This provides ~2x
+        # speed up when benchmarking fused AdamW on CPU
+        threads = os.cpu_count() // torch.distributed.get_world_size()
```
Comment (janeyx99) [inline recipes/full_finetune_distributed.py:590]: How did you calculate this? I opened the draftiest PR https://github.com/pytorch/pytorch/pull/126199 to account for other environments where CPU affinity is set and cgroups could be used, and plan on talking with Intel to get this into better shape. But why `os.cpu_count() // torch.distributed.get_world_size()`?

**Review by janeyx99 — APPROVED** [review]
Woo! Nice to integrate fused CPU adam into torchtune :)

File: recipes/full_finetune_distributed.py (hunk 2 above)
Comment (rohan-varma) [reply recipes/full_finetune_distributed.py:590]: @janeyx99 Unfortuantely I don't have the bets answer to this, @msaroufim suggested following this over chat and referred to answerAI doing similar (following his suggestions) - https://github.com/AnswerDotAI/fsdp_qlora/commit/1a9fddfb660b85e46bdf18c3fcf0d599628c9c6c

Intuitively, this makes sense since it's essentially assigning each process an equal number of vCPUs out of the vCPUs available. But as you mentioned, this doesn't take into account physical CPUs being < vCPUs, cgroups, and the like. Mostly meant to be a stopgap for now, and I'm sure perf will be very different on different HW set ups. In the medium term, we might want to use what you end up landing in torch core, or even expose this to the user so they can configure this.

Comment (msaroufim) [reply recipes/full_finetune_distributed.py:590]: Yeh I suspect comprehensive benchmarking with how diverse cpus are might take a while. So this is a fine stopgap

File: recipes/full_finetune_distributed.py [hunk 3]
```diff
@@ -103,6 +105,21 @@ def __init__(self, cfg: DictConfig) -> None:
...
+                    offloading, otherwise optimizer will use non vectorized kernels and cause massive slowdown.
+                    """
+                )
+            else:
+                # We are fusing + CPU offloading - check to make sure we're in a nightly.
+                if not utils.torch_version_ge("2.4.0"):
+                    raise RuntimeError(
+                        "Using fused optimizer on CPU is only supported in PyTorch nightly."
+                    )
+
```
Comment (kartikayk) [inline recipes/full_finetune_distributed.py:122]: I'm not a fan of this if-else block being in the main recipe at all. This is starting to get back into the territory of the wild-wild-west recipe scripts that we spoke about a while back. Let's please brainstorm a bit about how to handle this and come up with the right design for this.

File: recipes/full_finetune_distributed.py [hunk 4]
```diff
@@ -563,6 +584,12 @@ def recipe_main(cfg: DictConfig) -> None:
         )
 
     init_process_group(backend="gloo" if cfg.device == "cpu" else "nccl")
+    if cfg.get("fsdp_cpu_offload", False):
```
Comment (kartikayk) [inline recipes/full_finetune_distributed.py:587]: Again not a fan of this being in the recipe code, this should likely be in a utility or something which the user doesn't probably need to reason about.

cc: @ebsmothers to help brainstorm this a bit

**Review by kartikayk — CHANGES_REQUESTED** [review]
Thanks for adding this change! I think there are some open recipe UX questions here which I'd like for us to think through before we land this.

File: recipes/full_finetune_distributed.py (hunk 3 above)
Comment (ebsmothers) [reply recipes/full_finetune_distributed.py:122]: I gave a suggestion in _recipe_registry.py, at least related to the versioning side of things (but it's a bigger change than what we're looking at here). Separately we are kinda drawing a line (I don't think?) we've drawn before.. we are putting an explicit nightly check in one of our core 4 recipes. While I 

As for the actual validation of fsdp_cpu_offload + optimizer.fused, I guess it's a bit tricky to put in a single util since it's a more cross-cutting config pair. The only alternative I can think of offhand is to move into a config validation, which I think we also had legitimate concerns about doing previously. (I guess we can't validate directly on the optimizer either since it gets instantiated after the FSDP wrapping)

File: recipes/full_finetune_distributed.py (hunk 4 above)
Comment (ebsmothers) [reply recipes/full_finetune_distributed.py:587]: Yeah +1 can we put this in a utility?

File: recipes/configs/llama3/70B_full.yaml [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+# Fine-tuning arguments
+batch_size: 2
+epochs: 3
+
+optimizer:
+  _component_: torch.optim.AdamW
+  lr: 2e-5
+  foreach: False
+  fused: True
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply recipes/configs/llama3/70B_full.yaml:83]: This is a speedup only change (will clarify the amount of speed up we get once I re-run some benchmark)

File: recipes/full_finetune_distributed.py (hunk 3 above)
Comment (rohan-varma) [reply recipes/full_finetune_distributed.py:122]: Versioning recipes make sense, though this portion has a bit more than versioning since I also want to warn if fused optimizer is not used with CPU offload. 

> we are putting an explicit nightly check in one of our core 4 recipes.

We are - but only under a very specific case - CPU offload with FSDP distributed training.

Can definitely add a helper function, though let me know if we should consider something else / alternative design. @ebsmothers @kartikayk

File: recipes/full_finetune_distributed.py (hunk 4 above)
Comment (rohan-varma) [reply recipes/full_finetune_distributed.py:587]: Yeah, definitely!

**Comment by ebsmothers** [comment]
Do we know the story behind the large delta between peak active and reserved? (Basically I am wondering if this is similar to other cases where we saw that we could get peak reserved down just by constraining the available memory a bit. And if so, could we actually run this on 4x A100? That'd be nice)

File: recipes/full_finetune_distributed.py (hunk 4 above)
Comment (ebsmothers) [reply recipes/full_finetune_distributed.py:587]: Bumping this comment. Also I don't fully understand the code comment just below: do we only need to `set_num_threads` in the case of fused optimizer? Looking at the answer.ai code seems like the answer is no, but based on the code comment it's not clear to me

Comment (rohan-varma) [reply recipes/full_finetune_distributed.py:587]: We see the biggest gain setting this when fused optimizer is used as it controls threads used for intra-op parallelism, and CPU optimizer is the heaviest CPU op. We may have very slight speed ups when not using fused optimizer, but those can be investigated separately of this PR.

File: torchtune/utils/_distributed.py [hunk 6]
```diff
@@ -117,6 +117,22 @@ def init_distributed(**kwargs: Dict) -> bool:  # noqa: DOC106, DOC109
...
+    to be the number of physical CPU cores divided by the number of GPUs as we
+    use one process per GPU, and this avoids CPU oversubscription. Note that this is
+    currently a rough approximation, and doesn't take into account environments where
+    things like CPU affinity is set.
+    """
+    num_threads = os.cpu_count() // (
+        torch.distributed.get_world_size() if torch.distributed.is_initialized() else 1
+    )
+    torch.set_num_threads(num_threads)
+    log.info(f"Set intra op parallelism no. of threads to {num_threads}")
```
Comment (ebsmothers) [inline torchtune/utils/_distributed.py:133]: For some reason the logger in this file is `_log` and not `log`, so this errors out

**Review by ebsmothers — APPROVED** [review]
Thanks for working through all the comments here! This is awesome functionality to have and I like the way the UX turned out here. One more comment to address before merging but otherwise looks great!

File: torchtune/utils/_distributed.py (hunk 6 above)
Comment (rohan-varma) [reply torchtune/utils/_distributed.py:133]: Ouch, great catch

**Comment by andyl98** [comment]
Hi @rohan-varma , I was trying some fine-tuning runs with 8x A100-80G setup & 1T memory. The training loop works fine but when entering the checkpoint saving stage, the memory utilization continues to go up until it exceeds my node's limit. Do you know if this is a bug or an intended behavior? Thanks a lot for helping out!
Screenshot 2024-06-05 at 9 13 07 PM

**Comment by andyl98** [comment]
Update: I think I can successfully save the checkpoint after training if I don't checkpoint the optimizer states. Perhaps an option should be added to save memory by not saving optim states?

Screenshot 2024-06-06 at 2 13 20 PM

**Comment by IvanLukianenko** [comment]
Hi @andyl98, your code changes fix the problem in your setup and you can now save checkpoints at the end of epochs? I ran into the same problem in almost the same setup, but your code changes didn't help me. Many thanks for the help!

**Comment by musabgultekin** [comment]
@andyl98 Extra note: I was able to train&save with 4xA100 80GB with 1TB DRAM. It also worked with 8xA100 with 2TB DRAM. Haven't done 8xA100 with 1TB though like in your case, it might be that 1TB DRAM is too small for 8x.

**Comment by andyl98** [comment]
> @andyl98 Extra note: I was able to train&save with 4xA100 80GB with 1TB DRAM. It also worked with 8xA100 with 2TB DRAM. Haven't done 8xA100 with 1TB though like in your case, it might be that 1TB DRAM is too small for 8x.

Right, seems like the model full shard checkpointing stage inevitably consumes quite some memory. Appreciate the note!

**Comment by ebsmothers** [comment]
@andyl98 @musabgultekin sorry for the delay here, I agree it's suspicious that we'd require 1TB to save the 70B model and optimizer states. I created #1092 so we can track it there
